src/DataProtection/samples/KeyManagementSimulator/Program.cs

```diff
@@ -0,0 +1,591 @@
+        var authenticatedEncryptorConfiguration = new MockAlgorithmConfiguration(descriptor);
+
+        // Any xml element deserializes to the only descriptor
+        var descriptorDeserializer = new MockAuthenticatedEncryptorDescriptorDeserializer(descriptor);
+
+        // We control deserialization by hooking activation
+        var activator = new MockActivator(decryptor, descriptorDeserializer);
+
+        // We don't actually want to burn CPU by doing real encryption
+        var authenticatedEncryptor = new MockAuthenticatedEncryptor();
+
+        var authenticatedEncryptorFactory = new MockAuthenticatedEncryptorFactory(authenticatedEncryptor);
+
+        var keyManagementOptions = Options.Create(new KeyManagementOptions()
+        {
+            AuthenticatedEncryptorConfiguration = authenticatedEncryptorConfiguration,
+            XmlRepository = repository,
+            XmlEncryptor = encryptor,
+        });
+        keyManagementOptions.Value.AuthenticatedEncryptorFactories.Add(authenticatedEncryptorFactory);
+
+        var keyManager = new XmlKeyManager(keyManagementOptions, activator)
+        {
+            GetUtcNow = () => _now,
+        };
+        var defaultKeyResolver = new DefaultKeyResolver(keyManagementOptions);
+
+        _cacheableKeyRingProvider = new KeyRingProvider(keyManager, keyManagementOptions, defaultKeyResolver)
+        {
+            JitterRandom = productRandom,
+        };
+    }
+
+    /// <summary>
+    /// The problem-time-score for this instance.
+    /// 
+    /// This score is a little hard to explain.  Suppose there were two instances, A and B.
+    /// If for a period of length T, B had a default key that was unknown to A, then T would
+    /// be added to A's score.  So, if A didn't know about B's default key for 5 minutes at
+    /// the start of the run and another 3 hours later on (probably a different default key
+    /// by then), then A's score would be 3:05.  If there were a third instance, C, and
+    /// A didn't know about C's default key for 2:03, then A's total score would be 5:08,
+    /// regardless of whether or not B and C had any overlap in their missing key ranges.
+    /// This reflects the fact that not knowing the key of either B or C is *worse* than
+    /// just not knowing the key of one of them because it increases the likelihood of a
+    /// session migrating to A and hitting a missing key error.
+    /// </summary>
+    public TimeSpan CumulativeProblemTime => _cumulativeProblemTime;
+
+    /// <summary>
+    /// Trigger a key ring refresh and return enough information to update metrics
+    /// and advance the simulation.
+    /// </summary>
+    public KeyRingDescriptor RefreshKeyRingAndGetDescriptor(DateTimeOffset now)
+    {
+        // Update this before calling GetCacheableKeyRing so that it's available to XmlKeyManager
+        _now = now;
+
+        //Console.WriteLine($"[{_instanceNumber}] RefreshKeyRingAndGetDescriptor({now})");
```

Comment:
Keeping these? Maybe have a debug flag you can pass in which turns on logging?

Reply:
My thinking was that I was providing a sample of what one might want to log and that building in a verbosity system would be overkill.  I can get rid of them if you find them unhelpful.  Thoughts?

Reply:
Not worth revisiting - I've deleted those lines.